### PR TITLE
[No QA] Remove throw in registerNetworkResponseHandler. Clean up IOU actions file.

### PR DESCRIPTION
**Pull request by marcaaron**
cc @Julesssss 

### Details

I noticed that for some reason we are throwing an error from deep inside the API lib. These errors should not be thrown and caught with a `.catch()`. 

Seeing as we are going to re-use some of the APIs around IOU Reports for workspace chats I'd figure that doing a some cleanup on the actions files couldn't hurt.

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/201109

### Tests

1. Create Request (`CreateIOUTransaction`) + Split (`CreateIOUSplit`) Verify that IOU transactions can be created without issue
2. Force different `jsonCode` to be returned by these API when calling the methods (edit Web-Expensify api.php code and try to send `405`, `404`, and `402` verify the correct errors show)

Used this diff
```diff
diff --git a/api.php b/api.php
index 6386ea1afd8..8cc96460e5d 100755
--- a/api.php
+++ b/api.php
@@ -2633,6 +2633,7 @@ try {
         $response = TransactionAPI::createIOUTransaction($authToken, $_REQUEST['debtorEmail'] ?? '', $_REQUEST['amount'] ?? 0, $_REQUEST['currency'] ?? '', $_REQUEST['comment'] ?? '');
     } elseif ($command === 'CreateIOUSplit') {
         $response = ReportAPI::createIOUSplit($authToken, $_REQUEST['reportID'] ?? 0, $_REQUEST['amount'] ?? 0, $_REQUEST['splits'] ?? '', $_REQUEST['currency'] ?? '', $_REQUEST['comment'] ?? '');
+        $response['jsonCode'] = 402;
     } elseif ($command === 'Payroll_Summary') {
         $csv = PayrollAPI::getSummary($authToken, $policyID, $_REQUEST['startDate'] ?? '', $_REQUEST['endDate'] ?? '');
         header('Content-Type: text/csv');
```

Verified the alerts were created

### QA

- No QA as it may be difficult to get these errors to throw 🤔 
- Carefully test creating Requests and Split Bill features for any regressions.

### CODE REVIEW

**Comment by marcaaron** [comment]
cc @mountiny if you are interested in the review here and the changes as they relate to the recent work you did

**Comment by Julesssss** [comment]
Is this still WIP?

File: src/libs/actions/IOU.js [hunk 1]
```diff
@@ -9,145 +9,175 @@ import Navigation from '../Navigation/Navigation';
...
+ * @param {Object[]} requestParams
+ * @returns {Object}
+ */
+function prepareChatAndIOUReports(response, requestParams) {
+    const reports = response.response;
+    const chatReportsToUpdate = {};
+    const iouReportsToUpdate = {};
+
+    _.each(reports, (reportData) => {
+        // First, the existing chat report needs updated with the details about the new IOU
```
Comment (mountiny) [inline src/libs/actions/IOU.js:25]:
```suggestion
        // First, the existing chat report needs to be updated with the details about the new IOU
```

File: src/libs/actions/IOU.js [hunk 2]
```diff
@@ -9,145 +9,175 @@ import Navigation from '../Navigation/Navigation';
...
+            const chatReportID = paramsForIOUReport.chatReportID;
+            const chatReportKey = `${ONYXKEYS.COLLECTION.REPORT}${chatReportID}`;
+            chatReportsToUpdate[chatReportKey] = {
+                iouReportID: reportData.reportID,
+                total: reportData.total,
+                stateNum: reportData.stateNum,
+                hasOutstandingIOU: true,
+            };
+
+            // Second, the IOU report needs updated with the new IOU details too
```
Comment (mountiny) [inline src/libs/actions/IOU.js:37]:
```suggestion
            // Second, the IOU report needs to be updated with the new IOU details too
```

File: src/libs/actions/IOU.js [hunk 3]
```diff
@@ -156,14 +186,13 @@ function createIOUSplit(params) {
...
-    Onyx.merge(ONYXKEYS.IOU, {loading: true, creatingIOUTransaction: true, error: false});
+    startLoadingAndResetError();
 
     API.CreateIOUSplit({
         ...params,
         splits: JSON.stringify(params.splits),
     })
-        .then(() => Onyx.merge(ONYXKEYS.IOU, {loading: false, creatingIOUTransaction: false}))
-        .catch(() => Onyx.merge(ONYXKEYS.IOU, {error: true}));
+        .then(() => Onyx.merge(ONYXKEYS.IOU, {loading: false, creatingIOUTransaction: false}));
```
Comment (mountiny) [inline src/libs/actions/IOU.js:195]: How is error here being handled?

**Review by mountiny — COMMENTED** [review]
Thanks for the ping! Left just a few questions, looks great!

Oh sorry if it was not ready for a review yet, havent noticed it is a draft 🤦

File: src/libs/actions/IOU.js (hunk 3 above)
Comment (marcaaron) [reply src/libs/actions/IOU.js:195]: Which error do we need to handle? Where's it at?

Comment (marcaaron) [reply src/libs/actions/IOU.js:195]: I guess these maybe? https://github.com/Expensify/Auth/blob/d3cf26e0330c3a4cf98c6ac3abf141a53f617f33/auth/command/CreateIOUSplit.cpp#L15-L18
Should hit the standard handler https://github.com/Expensify/Web-Expensify/blob/7ecf0d26981bcf715f551c58a21744750c06fc83/lib/Auth.php#L115

And because they are 404 and we removed the default handler that rejects the promise we must now do something?

**Review by Julesssss — DISMISSED** [review]
Nice refactor and clean up. We'll need some tests, but the changes look good to me so far

**Comment by marcaaron** [comment]
Actually, going to scale back these changes so they are easier to test. 

I think we are going to have to fix this problem bit by bit and be sure errors are handled before removing `catch()`. Use of `catch()` is more widespread than I realized.

Gonna start with the IOU stuff only for now.

File: src/libs/API.js [hunk 4]
```diff
@@ -179,13 +179,7 @@ Network.registerResponseHandler((queuedRequest, response) => {
         return;
     }
 
-    if (response.jsonCode === 405 || response.jsonCode === 404 || response.jsonCode === 402) {
-        // IOU Split & Request money transactions failed, due to invalid amount(405), unable to split(404), invalid phone number(402)
-        // It's a failure, so reject the queued request
-        queuedRequest.reject(response);
-        return;
-    }
```
Comment (marcaaron) [inline src/libs/API.js:187]: There may be consequences of removing this code for some API that are wrapped in a `catch()`, but it will be too much work to look at all API methods using a catch and find out whether it throws these code. I'm only going the fix the ones that are explicitly being handled.

File: src/libs/actions/IOU.js [hunk 5]
```diff
@@ -41,115 +47,131 @@ function getIOUReportsForNewTransaction(requestParams) {
...
                     iouReportsToUpdate[iouReportKey] = Report.getSimplifiedIOUReport(reportData, chatReportID);
                 }
             });
 
-            // Now, merge the updated objects into our store
             Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT, chatReportsToUpdate);
-            return Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT_IOUS, iouReportsToUpdate);
+            Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT_IOUS, iouReportsToUpdate);
         })
-        .catch(() => Onyx.merge(ONYXKEYS.IOU, {error: true}))
```
Comment (deetergp) [inline src/libs/actions/IOU.js:54]: Are we removing `.catch()`-es because the jsonCodes from the API are really the only thing we have to go on? Nothing's being thrown in the JS layer that could be caught?

Comment (marcaaron) [reply src/libs/actions/IOU.js:54]: Can of worms, but generally speaking there's only a couple reasons why you end up in a catch and they are:

**`fetch()` errors**



**force promise to be rejected by throwing an exception because you are doing something bad in a `.then()` or intentionally threw a `new Error()`**



For the fetch errors we should be handling them globally in the network layer here for persisted things like report comments

https://github.com/Expensify/App/blob/2969ed7bd0635a98d07520a1a3ba09c985fdb7c3/src/libs/Network.js#L71-L76

Any persisted `fetch()` requests should not ever be getting rejected they are just retried or dumped at some point.

For the ones that _aren't persisted_  (one's being modified in this PR) we could trigger the `onError()` callback here where the request will get retried for a while and then eventually we **do** throw an error

https://github.com/Expensify/App/blob/2969ed7bd0635a98d07520a1a3ba09c985fdb7c3/src/libs/Network.js#L236-L259

Which lands us here and rejects the promise (which would lead to an unhandled promise rejection unless we add a catch when we call the method)

https://github.com/Expensify/App/blob/2969ed7bd0635a98d07520a1a3ba09c985fdb7c3/src/libs/API.js#L193-L207

So a promise _could_ be rejected in theory. But **if we are offline we should not be making the request in the first place**. 

Having fun yet?

Technically, we could probably argue that in the current state of the app all API methods should be using a `catch()` because there's always a chance that you were online -> make a request -> go offline -> get a fetch error. But barring any accidental exceptions caused by bad code _the only error we should expect to see_ is the `CONST.ERROR.API_OFFLINE` error. 

If you look at the catches I want to remove here. We can no longer make that assumption because we added other reasons why these promises could be rejected (which I'm undoing) or we explicitly throw errors in the `.then()` which isn't really necessary and makes the reason behind the `.catch()` ambiguous. Was it a rare "offline" error or the one we explicitly throw? Do we bother to check?

Basically, this whole business is way more complicated than it has to be and the complexity has led to a lot of non standard error handling.

File: src/libs/actions/IOU.js [hunk 6]
```diff
@@ -190,16 +218,16 @@ function rejectTransaction({
     })
         .then((response) => {
             if (response.jsonCode !== 200) {
-                throw new Error(`${response.code} ${response.message}`);
+                Log.hmmm('Error rejecting transaction', {error: response.error});
```
Comment (deetergp) [inline src/libs/actions/IOU.js:221]: NAB: Should we be including the error codes in these log messages?

**Review by deetergp — APPROVED** [review]
Thanks for the explanation, it definitely cleared things up. I had one NAB comment, but otherwise looks good to me.

File: src/libs/actions/IOU.js (hunk 6 above)
Comment (marcaaron) [reply src/libs/actions/IOU.js:221]: Maybe.. I didn't think super hard about this one, but I feel like we should be able to see the request and what error was thrown along with this log line and piece together what happened. This log line felt overly cautious or possibly unnecessary but I wasn't sure.

I think in the future we could consider having a default log any time an action returns a non-200 jsonCode

Comment (deetergp) [reply src/libs/actions/IOU.js:221]: > I think in the future we could consider having a default log any time an action returns a non-200 jsonCode

It would certainly make debugging easier if we did.
